(Closes #2510) ArrayRange2Loop transformation no longer allows Character arrays by default.

LonelyCat124 opened the pull request:
This fixes the code in #2510. I only check the lhs for now (since I believe types on lhs and rhs should match, and so this is easier) but it could be changed to check the rhs as well if appropriate (though I think really its the resultant string that we care about ignoring this on).

Additionally, I added the `allow_strings` option for other use-cases. If I've missed anything or the reviewer feels anything else should be in this PR let me know.

For me this is ok for pycodestyle/pylint and I updated the pydoc - if any additional text should be in the main documentation I can add that also.

Test added as well - if I missed anything I'll address it when the bot tells me coverage is missing.

arporter commented:
> I only check the lhs for now (since I believe types on lhs and rhs should match, and so this is easier) but it could be changed to check the rhs as well if appropriate

Sometimes we won't be able to identify the type and will get `UnknownType`, in which case it would then be good to try and check the RHS too. An optimisation might also be to first check for a character literal on the RHS as that's quite a common situation and faster for us to identify.

arporter commented:
Yeah, a wildcard (or even named) import for the thing being assigned to will do the trick for testing. My other suggestion might be a premature optimisation but I had something like `isinstance(assign.rhs, Literal) and assign.rhs.datatype.xxx == yyy` in a script I was messing with.

LonelyCat124 commented:
I've not added the rhs optimsiation you suggested. Is it possible for the RHS of a fortran character assignment statement to have multiple children? In C you can do things like `array[0] = 'u' + 2;` - I guess something like `arr(:) = 'b' // '2'` or anything else with concatenation?

LonelyCat124 commented:
This implementation fails on a call related to #1799. I will handle this by just ignoring these for now, but make a note that if that issue is fixed we can improve it again.

sergisiso commented:
I submitted the integration test to see if this has any impact on NEMO

sergisiso reviewed (changes requested):
Thanks for this @LonelyCat124 this could potentially fix some issues in NEMO too, but unfortunately we currently have 2 duplicated ArrayRange2Loop transformations, the one this PR modifies and the one under: src/psyclone/domain/nemo/transformations/nemo_arrayrange2loop_trans.py

The generic one is not good at correctly handling complex cases, so NEMO (and Socrates if it uses the normalise method) call the specialised NEMO one. It is a bit odd because the specialised one does not have any domain logic anymore, and should repace the generic one in #2004 . Until this is done, you will need to add the "allow_strings" options in both (or alternatively wait until #2004 if this PR is not urgent).

There is also the NemoAllArrayRange2LoopTrans and NemoOuterArrayRange2LoopTrans, that internally call the NemoArrayRange2Loop, but these don't pass the options to the inner call. You will need to fix all this internal apply and validate calls to make sure that the options are passed down.

LonelyCat124 commented on src/psyclone/psyir/transformations/arrayrange2loop_trans.py:
```diff
@@ -134,18 +135,26 @@ def validate(self, node, options=None):
 
         :param node: the node that is being checked.
         :type node: :py:class:`psyclone.psyir.nodes.Assignment`
+        :param options: a dictionary with options for transformations
+        :type options: Optional[Dict[str, Any]]
+        :param bool options["allow_string"]: whether to allow the
+                                             transformation on a character
+                                             type array range.
```
Added

LonelyCat124 commented on src/psyclone/psyir/transformations/arrayrange2loop_trans.py:
```diff
@@ -214,6 +223,40 @@ def validate(self, node, options=None):
+        except NotImplementedError:
+            pass
+        for child in node.rhs.walk((Literal, Reference)):
+            # Skip over indices
+            if child.ancestor(ArrayReference) is not None:
+                continue
```
The `test_validate_function_in_index` test shows why this is necessary. If you remove this line the code will fail. I think this was added because something in the exampes/transform stuff was failing because there was a Literal/Reference node in the rhs inside the array reference's we care about, e.g.
`b(:) = a(LBOUND(a):)`. I'm also unclear as to why this test doesn't fail anyway, since it should look at a and b but I assume because of whatever is wrong with the issue I say.

If you remove these lines then building in here fails:
`make[2]: Leaving directory '/home/aidan/PSyclone/examples/lfric/eg2'`

I think the example is why this code exists - I think because you can have unknownFortranType indexes or in Intrinsics and then it throws an exception.

LonelyCat124 replied:
I think it fails on a different exception - not a NotImplementedError but one where either the .datatype or .intrinsic doesn't exist on the symbol.

LonelyCat124 commented:
@sergisiso should be good for another look now

sergisiso commented on src/psyclone/domain/nemo/transformations/nemo_arrayrange2loop_trans.py:
```diff
@@ -306,6 +316,44 @@ def validate(self, node, options=None):
+            if child.ancestor(ArrayReference) is not None:
+                continue
+            try:
+                # Skip unresolved types
+                if isinstance(child.datatype, UnresolvedType):
+                    continue
```
Would adding this allows us to drop
```
            # Skip over indices
            if child.ancestor(ArrayReference) is not None:
                continue
```
from above?

sergisiso commented on src/psyclone/psyir/transformations/arrayrange2loop_trans.py:
```diff
@@ -214,6 +227,40 @@ def validate(self, node, options=None):
+            if child.ancestor(ArrayReference) is not None:
+                continue
+            try:
+                if child.datatype.intrinsic == ScalarType.Intrinsic.CHARACTER:
```
Some considerations (skip Unresolved, Unsupported and drop previous continue condition) as the NEMO version of the loop also apply here.

LonelyCat124 replied:
Changed

sergisiso reviewed (changes requested):
Thanks @LonelyCat124 for adding the option to the NEMO version of the transformations, which is what NEMO and Socrates use.

We are still missing to pass some option parameters down to apply and validates, see inline comment. But importantly NEMO/Socrates do 3 hops `normalise()->nemo_allarrayrange2loop_trans.py->nemo_outerarrayrange2loop_trans.py->nemo_arrayrange2loop_trans.py` and we are still missing the options in the outerarrayrange2loop call.

For this reason, I think it would be good to replicate the test from nemo_arrayrange2loop::test_character_validation into the nemo_allarrayrange2loop (which applies to the statement instead of the Range) and see that calling it with option False and True makes it all the way down.

LonelyCat124 commented on src/psyclone/domain/nemo/transformations/nemo_outerarrayrange2loop_trans.py:
```diff
@@ -108,6 +108,9 @@ def apply(self, node, options=None):
+                                             type array range.
 
         '''
         self.validate(node)
```
Added

LonelyCat124 commented on src/psyclone/domain/nemo/transformations/nemo_arrayrange2loop_trans.py:
```diff
@@ -108,9 +109,12 @@ def apply(self, node, options=None):
             transformation. This is an optional argument that defaults \
             to None.
         :type options: Optional[Dict[str, Any]]
+        :param bool options["allow_string"]: whether to allow the
+                                             transformation on a character
+                                             type array range.
```
Fixed.

LonelyCat124 commented on src/psyclone/psyir/transformations/arrayrange2loop_trans.py:
```diff
@@ -214,6 +227,40 @@ def validate(self, node, options=None):
 
+        if not options:
+            options = {}
+        allow_string_array = options.get("allow_string", False)
+        # If we allow string arrays then we can skip the check.
+        if allow_string_array:
+            return
```
Done

LonelyCat124 commented on src/psyclone/tests/domain/nemo/transformations/nemo_arrayrange2loop_trans_test.py:
```diff
@@ -617,3 +617,81 @@ def test_validate_not_outermost_range():
+
+    code = '''subroutine test()
+    character :: a(100)
+    character :: b(100)
```
Added this as a separate check because I think it is a little cleaner.

arporter replied:
Yeah - we don't support `LEN=xxx`. (We only support single characters I think - I can't remember why we have this limitation but it's probably because we included CHARACTER type in the PSyIR type system but then didn't actually need it for anything.)

LonelyCat124 commented:
I think this is ready for another look now sergi.
I removed the tests with `character(LEN=xxx)` - if we have an issue for that lack of support I could add an xfailing test for it as well but I'm not sure if we do.

sergisiso commented on src/psyclone/domain/nemo/transformations/nemo_arrayrange2loop_trans.py:
```diff
@@ -306,6 +315,42 @@ def validate(self, node, options=None):
+            # ArrayMixin datatype lookup can fail if the indices contain a
+            # Call or Intrinsic Call. We catch this exception and continue
+            # for now - TODO #1799
+            try:
+                if lhs.datatype.intrinsic == ScalarType.Intrinsic.CHARACTER:
+                    raise TransformationError(
+                        "The NemoArrayRange2LoopTrans transformation doesn't "
+                        "allow character arrays by default. This can be "
+                        "enabled by passing the allow_string option to the "
+                        "transformation."
+                    )
+            except NotImplementedError:
+                pass
+            for child in assignment.rhs.walk((Literal, Reference)):
+                try:
+                    # Skip unresolved types
+                    if isinstance(child.datatype,
+                                  (UnresolvedType, UnsupportedType)):
+                        continue
+                    if (child.datatype.intrinsic ==
+                            ScalarType.Intrinsic.CHARACTER):
+                        raise TransformationError(
+                            "The NemoArrayRange2LoopTrans transformation "
+                            "doesn't allow character arrays by default. This "
+                            "can be enabled by passing the allow_string "
+                            "option to the transformation."
+                        )
```
Nemo still fails at line 328 with:
```
psyclone/domain/nemo/transformations/nemo_arrayrange2loop_trans.py", line 328, in validate
    if lhs.datatype.intrinsic == ScalarType.Intrinsic.CHARACTER:
       ^^^^^^^^^^^^^^^^^^^^^^
    AttributeError: 'UnsupportedFortranType' object has no attribute 'intrinsic'
```
This is because that check is not preceded with the:
```
# Skip unresolved types
if isinstance(child.datatype, (UnresolvedType, UnsupportedType)):
    continue
```
as the right-hand-side check. If both need to be the same, could we merge both together by doing the loop for `for child in assignment.walk((Literal, Reference)):` and remove the top try?

Then, I see we let strings inside UnsupportedTypes pass the validation. This is fine (for now) but could you mention that in the docstring and include again the `character(LEN=100)`-style test. I would like to have what happens in this case clarified in docstring and tests because it is common in NEMO.

sergisiso reviewed (changes requested):
@LonelyCat124 I have not made a full review, but the integration tests for NEMO still fails, can you look at the comment below.

arporter commented:
The issue associated with this problem is #2448.

LonelyCat124 commented:
@sergisiso Added now, should be ok, tests still pass but codecov seems to not work for some reason.

sergisiso commented on src/psyclone/psyir/nodes/array_reference.py:
```diff
@@ -187,6 +187,12 @@ def datatype(self):
+            # If we have an array reference to a symbol that is not considered
+            # an array by PSyIR, we don't know how to retrieve its element type
+            # (TODO #2448)
+            return UnresolvedType()
```
This line is missing from CodeCov (what fails is the Github connector but you can see the results at https://app.codecov.io/gh/stfc/PSyclone/pull/2511).

The ArrayReference create method will not let you build such a reference, but you can take one of the existing references   in `src/psyclone/tests/psyir/nodes/array_reference_test.py::test_array_datatype`, replace its .symbol to a new DataSymbol of StructureType, and call the .datatype again to test it.

sergisiso commented on src/psyclone/psyir/transformations/arrayrange2loop_trans.py:
```diff
@@ -214,6 +238,31 @@ def validate(self, node, options=None):
+                    # Skip unresolved types
+                    if isinstance(child.datatype,
+                                  (UnresolvedType, UnsupportedType)):
+                        continue
```
This line is also missing from the coverage report, the same line for NEMO is not missing so hopefuly is only a matter to copy over one of the test cases to the generic version. (keep it the same if possible as this have to be merged soon into a single implementation and will be easier to do if we know exactly what is duplicated)

LonelyCat124 replied:
I'm very unclear which test is covering this for the Nemo version.

I expected it to be the `character(LEN=x` test but this still gives a `child.datatype` of ArrayType I think which is not causing the `continue` to be taken, This actually comes from a test structure of ararys apparently...

LonelyCat124 replied:
Which makes me feel this code isn't doing what we want it to do. I think we need a longer if statement:
```

LonelyCat124 replied:
```                   
 if (isinstance(child.datatype,
                                   (UnresolvedType, UnsupportedType)) or
                        (isinstance(child.datatype, ArrayType) and
                         isinstance(child.datatype.datatype,
                             (UnresolvedType, UnsupportedType))))
```

sergisiso reviewed (changes requested):
@LonelyCat124  Everything looks good now and the integration test results are all correct (GPU was slow after last commit but I assume just because Glados was busy). The only remaining part is to cover the 2 missing Codecov lines (see comments).

LonelyCat124 commented:
@sergisiso I thik I corrected/fixed missing coverage. codecov seemed to be completely stuck when I last tried, but I'm pretty sure the lines that were missing before are now tested or I changed them to do what we intended and they're covered.

sergisiso reviewed (approved):
Thanks for resolving all issues @LonelyCat124 . This now passes all coverage and all integration tests. It is ready to merge.
